### ISSUE TITLE
An object is serialized as an array and can't be deserialized

### ISSUE THREAD

**Issue by murtymaganti** [JamesNK/Newtonsoft.Json#2906]
When deserializing an object, getting the error '**Cannot create and populate list type Microsoft.Reporting.WebForms.ReportHierarchy**'. ReportHierarchy is part of Microsoft report viewer and this class extends Stack<Reportinfo> so when serializing it, it gets serialized an array instead of as an object.  On stack overflow I see suggestions to add JsonObject attrobute to the class but this class is in a third party library and I have no access to the class. What is the possible solution to fix this issue? Is there anything that can be set to JsonSerializationSettings instead of an attribute?

Thanks
Murty

**Comment by murtymaganti** [comment]
Thanks @elgonzo 

It is weird that Microsoft.Reporting.WebForms.ReportHierarchy class is not documented any where online but it is part of the Microsoft reporting viewer. 

Now I am trying to do pretty much what you suggested i.e. create an instance of ReportHeirarchy and populate all members from Json text. But that seems to be not possible either as that class doesn't have a parameter less constructor (public or private) so Activator.CreateInstance throws exception (I have to use reflection because it is not part of the referenced  DLL but report viewer creates it runtime and saves it to session). Wonder how Newtonsoft library is creating instances without parameter less constructors.

**Comment by murtymaganti** [comment]
I wish there is a way to pass 'JsonObject' attribute through JsonSerializationSettings in addition to placing the attribute on the class. That will solve this problem.

**Comment by elgonzo** [comment]
> I wish there is a way to pass 'JsonObject' attribute through JsonSerializationSettings in addition to placing the attribute on the class. That will solve this problem.

There is a way to achieve something similar by creating a custom contract resolver (by deriving from [`DefaultContractResolver`](https://www.newtonsoft.com/json/help/html/M_Newtonsoft_Json_Serialization_DefaultContractResolver_CreateObjectContract.htm)) that would create and set up a JsonObjectContract for `ReportHierarchy` instead of an JsonArrayContract.

But quite likely, this wouldn't get you far. I don't know the `ReportHierarchy` implementation, but it being a stack as you said means it can contain stack items that perhaps need to be part of the serialized data. And i suspect these stack items it holds by virtue of being a stack aren't represented by fields or properties of `ReportHierarchy`. And if that's the case, instructing Newtonsoft.Json to treat `ReportHierarchy` as a json object would then ultimately mean that Newtonsoft.Json would treat `ReportHierarchy` _not_ as a stack/collection, with the result that the stack items would be missing in the json.

**Comment by murtymaganti** [comment]
ReportHeirarchy doesn't have it's own state other than being a stack. As you pointed out, serializing the stack would lose the data as the data is held in it's private fields. I need to somehow apply MemberSerialization.Feilds to it so it saves the data.

I am aware that Activator.CreateInstance has overloads that take constructor parameters but I can't or don't want to call those. If the internal state is restored back during deserialization then there is no need for calling the constructor. Similar to a binary formatter serializer which simply saves the 'state' of an instance and restores back.

What I am doing here is part of moving our AspNet sessions to an external storage like Redis. I am not trying to serialize the report control, of course we shouldn't save controls or pages to the session. But the report controller internally saves the data to session (probably to remember current page, current report etc.). Unfortunately we are using a very old report viewer library which was probably designed for binary serialization but not Json.

Now, I am trying outside the application using the simple class below to see if it can be done. Let me know how (if possible) the below can be serialized and deserialized back as Dummy object. If this works, most likely ReportHeirarchy also will work. Appreciate and 
thank you for all your suggestions so far.

public class Dummy : Stack<string>
    {
        public Dummy(string name)
        {
            this.Push(name);
        }
   }

Dummy dummy = new Dummy("Newtonsoft");
string str = JsonConvert.SerializeObject(dummy);
dummy = JsonConvert.DeserializeObject(str);

**Comment by elgonzo** [comment]
> ReportHeirarchy doesn't have it's own state other than being a stack. As you pointed out, serializing the stack would lose the data as the data is held in it's private fields. I need to somehow apply MemberSerialization.Feilds to it so it saves the data.

No, that's not what i was saying. I am saying you trying to serialize an object that is a stack as a JsonObject will lead to the stack items not being serialized.

> Let me know how (if possible) the below can be serialized and deserialized back as Dummy object (considering you can't modify the class) 

Due to the parameterized `public Dummy(string name)` constructor, you can't deserialize it from a json array. So, you would need to use a custom JsonConverter for this that will invoke the parameterized constructor with the appropriate string value.

> I am aware that Activator.CreateInstance has overloads that take constructor parameters but I can't or don't want to call those. If the internal state is restored back during deserialization then there is no need for calling the constructor. Similar to a binary formatter serializer which simply saves the 'state' of an instance and restores back.

But you are not using the old binary serialization mechanism here anymore. Don't equate a json serializer to .NET's old and deprecated binary serialization mechanism. So, if you don't have an `ReportHierarchy` instance existing already that the deserializer could populate with data, then you gotta create a `ReportHierarchy` **_somehow_**. If you don't want to invoke _any_ of the constructors `ReportHierarchy` has (regardless of whether it's public or private, parameterless or parameterized) then your only chance is to find some factory method in that library that would create `ReportHierarchy` instances for you. If there is no such factory methods to be found either, well, then you either have to swallow the pill regarding Activator.CreateInstance, or stop right here with your attempts of wanting to deserialize `ReportHierarchy` and think of an entirely different approach that's basically avoiding `ReportHierarchy` being part of the (de)serialized data as far as the json serializer is concerned.

**Comment by murtymaganti** [comment]
_No, that's not what i was saying. I am saying that treating an object that is a stack as a JsonObject during serialization will lead to the stack items not being serialized_. -> Stack items are serialized even when it is serialized as JsonObject. This can be done by adding MemberSerialization.Feilds to the JsonObject attribute of the class. Sadly it is not working when setting it through contract resolver. I have created a ContractResolver extending the DefaultContractResolver and override the CreateContract method and setting the MemberSerialization.Feilds. It just ignores it. It works if I set IgnoreSerializableAttribute to false but that gives another error as the assemblies are not trusted. So in the end, using a custom contract resolver is also not working.

```c#
        protected override JsonContract CreateContract(Type objectType)
        {
            if (objectType.FullName.Contains("ReportHeirarchy"))
            {
                JsonObjectContract jsonObjectContract =  base.CreateObjectContract(objectType);
                jsonObjectContract.MemberSerialization = Newtonsoft.Json.MemberSerialization.Fields;
                return jsonObjectContract;
            }
            return base.CreateContract(objectType);
}
```

**Comment by elgonzo** [comment]
> Stack items are serialized even when it is serialized as JsonObject. This can be done by adding MemberSerialization.Feilds to the JsonObject attribute of the class. 

Ah, okay. So there is some ReportHierarchy field through which the stack items are accessible. I missed that earlier. My apologies...

**Comment by murtymaganti** [comment]
@elgonzo 
I think I can make it work somehow by setting 'IgnoreSerializableAttribute' to false. But running into a trust issue and getting error '[**....] implements ISerializable but cannot be serialized using the ISerializable interface because the current application is not fully trusted and ISerializable can expose secure data**'

Do you know how to address this issue? At runtime I can see my AppDomain is fully trusted but still getting the error.

**Comment by elgonzo** [comment]
> Do you know how to address this issue?

No, unfortunately not. :-(
